File: backend/pkg/database/repository/organization.go
package repository

import (
	"context"
	"encoding/json"

	"backend/pkg/core"
	"backend/pkg/database/models"
	"backend/pkg/types"

	"github.com/google/uuid"
	"gorm.io/datatypes"
	"gorm.io/gorm"
)

// OrganizationRepository defines the interface for organization data access
type OrganizationRepository interface {
	FindByID(ctx context.Context, id uuid.UUID) (*types.Organization, error)
	FindBySlug(ctx context.Context, slug string) (*types.Organization, error)
	List(ctx context.Context) ([]types.Organization, error)
	Save(ctx context.Context, org *types.Organization) error
	Update(ctx context.Context, org *types.Organization) error
	Delete(ctx context.Context, id uuid.UUID) error
}

// organizationRepository provides data access for organizations
type organizationRepository struct {
	db *gorm.DB
}

// NewOrganizationRepository creates a new OrganizationRepository
func NewOrganizationRepository(db *gorm.DB) OrganizationRepository {
	return &organizationRepository{db: db}
}

// orgModelToType converts a database model to types
func orgModelToType(m *models.Organization) *types.Organization {
	var socialLinks map[string]interface{}
	if m.SocialLinks != nil {
		_ = json.Unmarshal(m.SocialLinks, &socialLinks)
	}

	return &types.Organization{
		ID:              m.ID,
		Name:            m.Name,
		Slug:            m.Slug,
		Bio:             m.Bio,
		LogoURL:         m.LogoURL,
		WebsiteURL:      m.WebsiteURL,
		EmailPublic:     m.EmailPublic,
		SocialLinks:     socialLinks,
		MetaDescription: m.MetaDescription,
		CreatedAt:       m.CreatedAt,
		UpdatedAt:       m.UpdatedAt,
	}
}

// orgTypeToModel converts a types type to database model
func orgTypeToModel(o *types.Organization) *models.Organization {
	var socialLinks datatypes.JSON
	if o.SocialLinks != nil {
		data, _ := json.Marshal(o.SocialLinks)
		socialLinks = datatypes.JSON(data)
	}

	return &models.Organization{
		ID:              o.ID,
		Name:            o.Name,
		Slug:            o.Slug,
		Bio:             o.Bio,
		LogoURL:         o.LogoURL,
		WebsiteURL:      o.WebsiteURL,
		EmailPublic:     o.EmailPublic,
		SocialLinks:     socialLinks,
		MetaDescription: o.MetaDescription,
		CreatedAt:       o.CreatedAt,
		UpdatedAt:       o.UpdatedAt,
	}
}

// FindByID retrieves an organization by its ID
func (r *organizationRepository) FindByID(ctx context.Context, id uuid.UUID) (*types.Organization, error) {
	var model models.Organization
	if err := r.db.WithContext(ctx).First(&model, id).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, core.ErrNotFound
		}
		return nil, err
	}
	return orgModelToType(&model), nil
}

// FindBySlug retrieves an organization by its slug
func (r *organizationRepository) FindBySlug(ctx context.Context, slug string) (*types.Organization, error) {
	var model models.Organization
	if err := r.db.WithContext(ctx).Where("slug = ?", slug).First(&model).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, core.ErrNotFound
		}
		return nil, err
	}
	return orgModelToType(&model), nil
}

// List retrieves all organizations
func (r *organizationRepository) List(ctx context.Context) ([]types.Organization, error) {
	var orgModels []models.Organization
	if err := r.db.WithContext(ctx).Order("created_at DESC").Find(&orgModels).Error; err != nil {
		return nil, err
	}

	orgs := make([]types.Organization, len(orgModels))
	for i, m := range orgModels {
		orgs[i] = *orgModelToType(&m)
	}
	return orgs, nil
}

// Save creates a new organization
func (r *organizationRepository) Save(ctx context.Context, org *types.Organization) error {
	model := orgTypeToModel(org)
	if model.ID == uuid.Nil {
		model.ID = uuid.New()
		org.ID = model.ID
	}
	return r.db.WithContext(ctx).Create(model).Error
}

// Update updates an existing organization
func (r *organizationRepository) Update(ctx context.Context, org *types.Organization) error {
	model := orgTypeToModel(org)
	return r.db.WithContext(ctx).Save(model).Error
}

// Delete removes an organization by its ID
func (r *organizationRepository) Delete(ctx context.Context, id uuid.UUID) error {
	result := r.db.WithContext(ctx).Delete(&models.Organization{}, id)
	if result.Error != nil {
		return result.Error
	}
	if result.RowsAffected == 0 {
		return core.ErrNotFound
	}
	return nil
}

File: backend/src/database/repository/organization.rs
use async_trait::async_trait;
use chrono::Utc;
use diesel::{
    ExpressionMethods, OptionalExtension, QueryDsl, SelectableHelper,
    result::{DatabaseErrorKind, Error as DieselError},
};
use diesel_async::RunQueryDsl;
use serde_json::{Map, Value};
use uuid::Uuid;

use crate::{
    core::organization::{Organization, OrganizationAccountRepository, OrganizationRepository},
    database::{
        models::organization::{NewOrganizationRow, OrganizationRow},
        pool::PgPool,
    },
    error::AppError,
    schema::{account, organization},
};

#[derive(Clone)]
pub struct DieselOrganizationRepository {
    pool: PgPool,
}

impl DieselOrganizationRepository {
    pub const fn new(pool: PgPool) -> Self {
        Self { pool }
    }

    async fn connection(
        &self,
    ) -> Result<
        diesel_async::pooled_connection::deadpool::Object<diesel_async::AsyncPgConnection>,
        AppError,
    > {
        self.pool.get().await.map_err(|_| AppError::Database)
    }
}

#[async_trait]
impl OrganizationRepository for DieselOrganizationRepository {
    async fn find_by_id(&self, id: Uuid) -> Result<Organization, AppError> {
        let mut connection = self.connection().await?;
        organization::table
            .find(id)
            .select(OrganizationRow::as_select())
            .first::<OrganizationRow>(&mut connection)
            .await
            .optional()
            .map_err(map_diesel_error)?
            .ok_or(AppError::NotFound)?
            .try_into()
    }

    async fn find_by_slug(&self, slug: &str) -> Result<Organization, AppError> {
        let mut connection = self.connection().await?;
        organization::table
            .filter(organization::slug.eq(slug))
            .select(OrganizationRow::as_select())
            .first::<OrganizationRow>(&mut connection)
            .await
            .optional()
            .map_err(map_diesel_error)?
            .ok_or(AppError::NotFound)?
            .try_into()
    }

    async fn list(&self) -> Result<Vec<Organization>, AppError> {
        let mut connection = self.connection().await?;
        let rows = organization::table
            .order(organization::created_at.desc())
            .select(OrganizationRow::as_select())
            .load::<OrganizationRow>(&mut connection)
            .await
            .map_err(map_diesel_error)?;
        rows.into_iter()
            .map(Organization::try_from)
            .collect::<Result<Vec<_>, _>>()
    }

    async fn save(&self, value: &mut Organization) -> Result<(), AppError> {
        if value.id.is_nil() {
            value.id = Uuid::new_v4();
        }
        let social_links = social_links_to_value(value);
        let mut connection = self.connection().await?;
        diesel::insert_into(organization::table)
            .values(NewOrganizationRow {
                id: value.id,
                name: &value.name,
                slug: &value.slug,
                bio: value.bio.as_deref(),
                logo_url: value.logo_url.as_deref(),
                website_url: value.website_url.as_deref(),
                email_public: value.email_public.as_deref(),
                social_links,
                meta_description: value.meta_description.as_deref(),
            })
            .execute(&mut connection)
            .await
            .map(|_| ())
            .map_err(map_diesel_error)
    }

    async fn update(&self, value: &Organization) -> Result<(), AppError> {
        let social_links = social_links_to_value(value);
        let mut connection = self.connection().await?;
        let affected = diesel::update(organization::table.find(value.id))
            .set((
                organization::name.eq(&value.name),
                organization::slug.eq(&value.slug),
                organization::bio.eq(value.bio.as_deref()),
                organization::logo_url.eq(value.logo_url.as_deref()),
                organization::website_url.eq(value.website_url.as_deref()),
                organization::email_public.eq(value.email_public.as_deref()),
                organization::social_links.eq(social_links),
                organization::meta_description.eq(value.meta_description.as_deref()),
                organization::updated_at.eq(Utc::now()),
            ))
            .execute(&mut connection)
            .await
            .map_err(map_diesel_error)?;
        if affected == 0 {
            Err(AppError::NotFound)
        } else {
            Ok(())
        }
    }

    async fn delete(&self, id: Uuid) -> Result<(), AppError> {
        let mut connection = self.connection().await?;
        let affected = diesel::delete(organization::table.find(id))
            .execute(&mut connection)
            .await
            .map_err(map_diesel_error)?;
        if affected == 0 {
            Err(AppError::NotFound)
        } else {
            Ok(())
        }
    }
}

#[async_trait]
impl OrganizationAccountRepository for DieselOrganizationRepository {
    async fn set_organization(
        &self,
        account_id: Uuid,
        organization_id: Option<Uuid>,
    ) -> Result<bool, AppError> {
        let mut connection = self.connection().await?;
        diesel::update(account::table.find(account_id))
            .set((
                account::organization_id.eq(organization_id),
                account::updated_at.eq(Utc::now()),
            ))
            .execute(&mut connection)
            .await
            .map(|affected| affected > 0)
            .map_err(map_diesel_error)
    }
}

fn social_links_to_value(organization: &Organization) -> Option<Value> {
    organization
        .social_links
        .clone()
        .map(|values| Value::Object(values.into_iter().collect::<Map<_, _>>()))
}

fn map_diesel_error(error: DieselError) -> AppError {
    match error {
        DieselError::NotFound => AppError::NotFound,
        DieselError::DatabaseError(DatabaseErrorKind::UniqueViolation, _) => {
            AppError::Conflict("resource already exists".to_owned())
        }
        _ => AppError::Database,
    }
}
